src/vcpkg/binarycaching.cpp

```diff
@@ -1094,222 +1069,141 @@ namespace
         }
-        const auto out = cmd_execute_and_capture_output(cmd);
-        if (out.exit_code == 0)
+
+        bool upload_file(StringView object, const Path& archive) const override
         {
-            return true;
+            auto cmd = command().string_arg("-q").string_arg("cp").string_arg(archive).string_arg(object);
+            const auto out = cmd_execute_and_capture_output(cmd);
+            if (out.exit_code == 0)
+            {
+                return true;
+            }
+
+            msg::println(Color::warning,
+                         msgObjectStorageToolFailed,
+                         msg::exit_code = out.exit_code,
+                         msg::tool_name = Tools::GSUTIL);
+            msg::write_unlocalized_text_to_stdout(Color::warning, out.output);
+            return false;
         }
 
-        msg::println(Color::warning, msgAwsFailedToDownload, msg::exit_code = out.exit_code);
-        msg::write_unlocalized_text_to_stdout(Color::warning, out.output);
-        return false;
-    }
+        bool download_file(StringView object, const Path& archive) const override
+        {
+            auto cmd = command().string_arg("-q").string_arg("cp").string_arg(object).string_arg(archive);
+            const auto out = cmd_execute_and_capture_output(cmd);
+            if (out.exit_code == 0)
+            {
+                return true;
+            }
+
+            msg::println(Color::warning,
+                         msgObjectStorageToolFailed,
+                         msg::exit_code = out.exit_code,
+                         msg::tool_name = Tools::GSUTIL);
+            msg::write_unlocalized_text_to_stdout(Color::warning, out.output);
+            return false;
+        }
+    };
 
-    struct AwsBinaryProvider : IBinaryProvider
+    struct AwsBinaryProvider : ObjectStorageProvider
     {
         AwsBinaryProvider(const VcpkgPaths& paths,
                           std::vector<std::string>&& read_prefixes,
                           std::vector<std::string>&& write_prefixes,
                           const bool no_sign_request)
-            : paths(paths)
-            , m_read_prefixes(std::move(read_prefixes))
-            , m_write_prefixes(std::move(write_prefixes))
+            : ObjectStorageProvider(paths, std::move(read_prefixes), std::move(write_prefixes))
             , m_no_sign_request(no_sign_request)
         {
         }
 
-        static std::string make_aws_path(const std::string& prefix, const std::string& abi)
-        {
-            return Strings::concat(prefix, abi, ".zip");
-        }
-
-        void prefetch(View<Dependencies::InstallPlanAction> actions, View<CacheStatus*> cache_status) const override
-        {
-            auto& fs = paths.get_filesystem();
+        StringLiteral vendor() const override { return "AWS"; }
 
-            const auto timer = ElapsedTimer::create_started();
+        Command command() const { return Command{paths.get_tool_exe(Tools::AWSCLI)}; }
```

Comment:
Should this include the `--no-sign-request` logic? That way it doesn't have to be repeated in every command.

Reply:
I'm not versed enough with the aws cli to know whether it accepts that option on every subcommand